tools/src/main/java/org/apache/kafka/tools/reassign/ReassignPartitionsCommand.java

```diff
@@ -0,0 +1,1501 @@
+import org.apache.kafka.server.util.json.JsonObject;
+import org.apache.kafka.server.util.json.JsonValue;
+import org.apache.kafka.tools.TerseException;
+import org.apache.kafka.tools.ToolsUtils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+@SuppressWarnings("ClassDataAbstractionCoupling")
+public class ReassignPartitionsCommand {
+    private static final String ANY_LOG_DIR = "any";
+
+    static final String HELP_TEXT = "This tool helps to move topic partitions between replicas.";
+
+    private static final DecodeJson.DecodeInteger INT = new DecodeJson.DecodeInteger();
+
+    private static final DecodeJson.DecodeString STRING = new DecodeJson.DecodeString();
+
+    private static final DecodeJson<List<Integer>> INT_LIST = DecodeJson.decodeList(INT);
+
+    private static final DecodeJson<List<String>> STRING_LIST = DecodeJson.decodeList(STRING);
+
+    /**
+     * The earliest version of the partition reassignment JSON.  We will default to this
+     * version if no other version number is given.
+     */
+    static final int EARLIEST_VERSION = 1;
+
+    /**
+     * The earliest version of the JSON for each partition reassignment topic.  We will
+     * default to this version if no other version number is given.
+     */
+    static final int EARLIEST_TOPICS_JSON_VERSION = 1;
+
+    // Throttles that are set at the level of an individual broker.
+    //DynamicConfig.Broker.LeaderReplicationThrottledRateProp
+    static final String BROKER_LEVEL_LEADER_THROTTLE = "leader.replication.throttled.rate";
+    //DynamicConfig.Broker.FollowerReplicationThrottledRateProp
+    static final String BROKER_LEVEL_FOLLOWER_THROTTLE = "follower.replication.throttled.rate";
+    //DynamicConfig.Broker.ReplicaAlterLogDirsIoMaxBytesPerSecondProp
+    static final String BROKER_LEVEL_LOG_DIR_THROTTLE = "replica.alter.log.dirs.io.max.bytes.per.second";
+    static final List<String> BROKER_LEVEL_THROTTLES = Arrays.asList(
+        BROKER_LEVEL_LEADER_THROTTLE,
+        BROKER_LEVEL_FOLLOWER_THROTTLE,
+        BROKER_LEVEL_LOG_DIR_THROTTLE
+    );
+
+    // Throttles that are set at the level of an individual topic.
+    //LogConfig.LEADER_REPLICATION_THROTTLED_REPLICAS_CONFIG
+    static final String TOPIC_LEVEL_LEADER_THROTTLE = "leader.replication.throttled.replicas";
+    //LogConfig.FOLLOWER_REPLICATION_THROTTLED_REPLICAS_CONFIG
+    static final String TOPIC_LEVEL_FOLLOWER_THROTTLE = "follower.replication.throttled.replicas";
```

Comment:
Is it a good idea to duplicate parameter's names constants?
What about moving `DynamicConfig` and `LogConfig` to `server-commons` project?
Of course not in the scope of this ticket. But duplicate parameter name constants are potentially dangerous.

Reply:
I think it's a gread idea. Let's discuss it on the dev-list.